smarts/env/wrappers/metrics.py

```diff
@@ -0,0 +1,284 @@
+    beginning with an underscore, to ensure security of the metrics computed.
+
+    Args:
+        env (gym.Env): A gym.Env to be wrapped.
+
+    Raises:
+        AttributeError: Upon accessing an attribute beginning with an underscore.
+
+    Returns:
+        gym.Env: A wrapped gym.Env which computes agents' performance metrics.
+    """
+
+    def __init__(self, env: gym.Env):
+        env = _Metrics(env)
+        super().__init__(env)
+
+
+class _Metrics(gym.Wrapper):
+    """Computes agents' performance metrics in a SMARTS environment."""
+
+    def __init__(self, env: gym.Env):
+        super().__init__(env)
+        _check_env(env)
+        self._cur_scen: str
+        self._cur_agents: Set[str]
+        self._steps: Dict[str, int]
+        self._done_agents: Set[str]
+        self._records = {}
+
+    def step(self, action: Dict[str, Any]):
+        """Steps the environment by one step."""
+        obs, rewards, dones, info = super().step(action)
+
+        # Only count steps in which an ego agent is present.
+        if len(obs) == 0:
+            return obs, rewards, dones, info
+
+        # fmt: off
+        for agent_name, agent_obs in obs.items():
+            self._steps[agent_name] += 1
+
+            # Compute all cost functions.
+            costs = Costs()
+            for field in fields(self._records[self._cur_scen][agent_name].cost_funcs):
+                cost_func = getattr(self._records[self._cur_scen][agent_name].cost_funcs, field.name)
+                new_costs = cost_func(agent_obs)
+                costs = _add_dataclass(new_costs, costs)
+
+            # Update stored costs.
+            self._records[self._cur_scen][agent_name].record.costs = costs
+
+            if dones[agent_name]:
+                self._done_agents.add(agent_name)
+                steps_adjusted, goals = 0, 0
+                if agent_obs.events.reached_goal:
+                    steps_adjusted = self._steps[agent_name]
+                    goals = 1
+                elif (
+                    len(agent_obs.events.collisions) > 0
+                    or agent_obs.events.off_road
+                    or agent_obs.events.reached_max_episode_steps
+                ):
+                    steps_adjusted = self.env.agent_specs[agent_name].interface.max_episode_steps
+                else:
+                    raise MetricsError(
+                        f"Unsupported agent done reason. Events: {agent_obs.events}."
+                    )
+
+                # Update stored counts.
+                counts = Counts(
+                    episodes=1,
+                    steps=self._steps[agent_name],
+                    steps_adjusted=steps_adjusted,
+                    goals=goals,
+                )
+                self._records[self._cur_scen][agent_name].record.counts = _add_dataclass(
+                    counts, 
+                    self._records[self._cur_scen][agent_name].record.counts
+                )
```

Comment:
I have mixed feelings about this approach. Though the immutability is nice, it seems like a bit of a burdensome way to maintain state. Creating a new dict and dataclass each step also seems inefficient. I don't think it's too critical as long as the performance is ok though.